File: library/src/main/java/com/github/pelmenstar1/rangecalendar/TextUtils.java
package com.github.pelmenstar1.rangecalendar;

import android.graphics.Paint;
import android.graphics.Rect;

import org.jetbrains.annotations.NotNull;

final class TextUtils {
    @NotNull
    private static final Rect textBoundsCache = new Rect();
    private static final Paint tempPaint = new Paint();

    private TextUtils() {
    }

    public static long getTextBounds(
            @NotNull String text,
            float textSize
    ) {
        tempPaint.setTextSize(textSize);

        return getTextBounds(text, tempPaint);
    }

    public static long getTextBounds(
            char @NotNull [] text,
            int offset, int length,
            float textSize
    ) {
        tempPaint.setTextSize(textSize);

        return getTextBounds(text, offset, length, tempPaint);
    }

    public static long getTextBounds(char @NotNull [] text, int index, int length, @NotNull Paint paint) {
        paint.getTextBounds(text, index, length, textBoundsCache);

        return packTextBoundsCache();
    }

    public static long getTextBounds(@NotNull String text, @NotNull Paint paint) {
        paint.getTextBounds(text, 0, text.length(), textBoundsCache);

        return packTextBoundsCache();
    }

    private static long packTextBoundsCache() {
        Rect bounds = textBoundsCache;

        return PackedSize.create(bounds.width(), bounds.height());
    }
}

File: library/src/main/java/io/github/pelmenstar1/rangecalendar/TextUtils.kt
package io.github.pelmenstar1.rangecalendar

import android.graphics.Paint
import android.graphics.Rect
import io.github.pelmenstar1.rangecalendar.PackedSize

private val textBoundsCache = Rect()
private val tempPaint = Paint()

fun getTextBounds(text: String, textSize: Float): Long {
    tempPaint.textSize = textSize

    return tempPaint.getTextBounds(text)
}

fun getTextBounds(text: CharArray, offset: Int, length: Int, textSize: Float): Long {
    tempPaint.textSize = textSize

    return tempPaint.getTextBounds(text, offset, length)
}

fun Paint.getTextBounds(text: CharArray, index: Int, length: Int): Long {
    getTextBounds(text, index, length, textBoundsCache)

    return packTextBoundsCache()
}

fun Paint.getTextBounds(text: String): Long {
    getTextBounds(text, 0, text.length, textBoundsCache)

    return packTextBoundsCache()
}

private fun packTextBoundsCache(): Long {
    val bounds = textBoundsCache

    return PackedSize.create(bounds.width(), bounds.height())
}